.devops/nix/package.nix

```diff
@@ -157,6 +163,8 @@ effectiveStdenv.mkDerivation (
       substituteInPlace ./*.py --replace "/usr/bin/env python" "${llama-python}/bin/python"
     '';
 
+    __noChroot = useMetalKit && effectiveStdenv.isDarwin;
```

Comment:
There's an undocumented feature [used in Nixpkgs](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/stdenv/generic/make-derivation.nix#L443-L448), called [`__impureHostDeps`](https://github.com/NixOS/nix/blob/5c8983b9e6bd47f1d921e21b717f936a02697344/src/libstore/build/local-derivation-goal.cc#L650-L652) which lets you demand specific paths from the host. That would be more fine-grained than disabling the sandbox.

That said, Nixpkgs seems to expose its own [`xcrun`](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L54-L106) as a passthru at `xcbuild.xcrun`: https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L118. Are you saying it's broken? If it is we should report it

Reply:
I don't think that `xcrun` shim will work here. Looking at source and testing on my system, [the `xcrun` shim from `xcbuild`](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L54-L106) expects whatever tool it is calling to be on `$PATH` so that the tool can be found with [`exec "$@"`](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L100C4-L100C13). This works for some binaries in `/usr/bin`, such as `cc` and `clang` (which are Apple-provided shims for the full binaries in `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/`.

However, the `metal` compiler (and the directory it's in, `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/metal`) are not added to $PATH because there may be multiple installs of Xcode on a system. `xcrun` handles running the `metal` binary from whichever Xcode installation is the default or provided via an environment variable. 

While investigating, it looks like the [`xcbuild` project from Facebook includes a more robust implementation of `xcrun`](https://github.com/facebookarchive/xcbuild/blob/dbaee552d2f13640773eb1ad3c79c0d2aca7229c/Libraries/xcsdk/Tools/xcrun.cpp) which can detect the install location of Xcode and run binaries from this location. From a brief glance at the code for that implementation, it looks through environment variables, the user's $HOME dir, and the standard install locations. 

That said, I don't think this ultimately solves the problem of needing an Xcode-specific binary to build llama.cpp on MacOS. Even with a shim for `xcrun`, the `metal` shader compiler from Xcode is still necessary. And if we're going to be calling Apple-provided binaries, I don't see any reason to avoid calling the normal `xcrun`, particularly as the [shell implementation](https://github.com/NixOS/nixpkgs/blob/2d5db19dffbb7fd86c8eba10a77deb8b8838ece3/pkgs/development/tools/xcbuild/wrapper.nix#L54-L106) only works for binaries already on $PATH, and the [`xcbuild` implementation from Facebook](https://github.com/facebookarchive/xcbuild/tree/master) is archived/ no longer maintained.

Reply:
@josephst did you try the shim? Xcrun here is only used for the build, inside the sandbox, and it'll use the toolchain provisioned by nix

Reply:
Even when using the `xcrun` binary provided by Xcode/Apple, the rest of the build is using the toolchain provided by Nix and the final `./result/bin/llama` binary confirms that it is built with clang 16.0.6, provided by Nix toolchain (rather than clang 15.0.0, which is the system default installed with Xcode). 

The problem I'm having with the `xcrun` shim is that it cannot locate the `metal` shader compiler. `metal` is unfortunately proprietary so there's no nix equivalent (unlike ie `clang`; `xcrun clang` will look for `clang` on $PATH and find it as part of the nix toolchain)


To illustrate the problem:

With the "full" `xcrun` provided by MacOS/Xcode:
```
$ /usr/bin/xcrun -sdk macosx metal --version
Apple metal version 32023.155 (metalfe-32023.155)
Target: air64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/metal/macos/bin
```

With the shim `xcrun`:
```
$ nix run nixpkgs#xcbuild.xcrun -- -sdk macosx metal --version
/nix/store/d0hnw5gy6mji9hqxcs564xx6zvba5m4s-xcrun/bin/xcrun: line 42: exec: metal: not found
```

Reply:
Ah, I just realized we haven't got an aarch64-darwin builder on github.

Could you please share the build logs? If your xcode/metal is from the host system and you had the sandbox enabled, how can the build access it?

Reply:
Hmm looks like I didn't have the sandbox properly enabled earlier. Just set `sandbox = true` and getting some new build errors. Lots of output, so [added both options (`__impureHostDeps` vs `noChroot`)  to a Gist](https://gist.github.com/josephst/38d30ffdd2543209c1206ec1dc6a7c43). 

As I work on this more, I'm starting to think that `__noChroot = true` is the better option. Trying to add `/usr/bin/xcrun` to `__impureHostDeps` with `sandbox=true` generates several new errors about how `/usr/bin/xcrun` isn't one of [the allowedImpureHostPrefixes](https://github.com/NixOS/nix/blob/7c506432abab84d79744f4454aa20fe0a458e0fb/src/libstore/globals.cc#L83). 

With `sandbox = true` in my `nix.conf`, the only way to get this to build is to go back to using `__noChroot = effectiveStdenv.isDarwin && useMetalKit;`, then building with `nix build .# --rebuild --option sandbox relaxed`. Sandboxing on macOS with Nix defaults to `false` so builds should work on macOS with `sandbox = off` or `sandbox = "relaxed"`.

Reply:
I see. Thanks for the research! The last bit I fail to understand is what has changed that it's now required to use the `metal` compiler from `xcode`: the aarch64-darwin always linked MetalKit and it worked at the time of https://github.com/ggerganov/llama.cpp/pull/4605#issuecomment-1869692179

CC @philiptaron

Reply:
Not exactly sure which commit changed it, but [looks like previously](https://github.com/ggerganov/llama.cpp/blob/58ba655af054715c0516ee270ad028ad9e74f357/CMakeLists.txt#L177) the `metal` compiler was only called if `LLAMA_METAL_SHADER_DEBUG` was true. 

Now, it's always called and `LLAMA_METAL_SHADER_DEBUG` affects which flags it's called with. Looks like the change was [about a week ago](https://github.com/ggerganov/llama.cpp/commit/381da2d9f0940d7009e3e918bed36338c8ff2fbb)